### PR TITLE
Release 12 h1 w4651 w4652 w4416

### PR DESCRIPTION
**Pull request by Neve**
This pull request fixes:
- https://wush.net/trac/rightscale/ticket/4651 (sys_firewall::setup_rule failes if protocol is set to ANY)
- https://wush.net/trac/rightscale/ticket/4652 (logging::default on reboot/stop_start duplicates same line if a remote server is given)
- https://wush.net/trac/rightscale/ticket/4416 (app_passenger/project/environment (Rails Environment) input should be drop down)

### CODE REVIEW

File: cookbooks/sys_firewall/recipes/setup_rule.rb [hunk 1]
```diff
@@ -15,6 +15,13 @@
 rule_protocol = node[:sys_firewall][:rule][:protocol]
 to_enable = (node[:sys_firewall][:rule][:enable] == "enable") ? true : false
 
+# iptables cannot create rule using "--protocol any" and "--dport"
+# "any" statement equals to TCP, UDP, and ICMP, see: http://goo.gl/16b6C
+# ICMP protocol do not use port numbers
+# so we will warn user to avoid iptables errors and blind configuration
+raise "Cannot create rule for protocol= #{rule_protocol} and port = #{rule_port} please create separate rules for each protocol " if (rule_protocol.include? "all") and (rule_port != nil)
```
Comment (douglaswth) [inline cookbooks/sys_firewall/recipes/setup_rule.rb]: Please revert this change and instead rename the `"all"` choice to `"both"` per my comment on the ticket.

File: cookbooks/logging_rsyslog/providers/default.rb [hunk 2]
```diff
@@ -60,13 +60,20 @@
...
+      if ::File.open('/etc/rsyslog.conf', 'r') { |f| f.read }.include? "#{remote_server}"
+        log "  #{remote_server} entry already exists. Skipping"
+      else
+        log "  Configuring Redhat/CentOS."
+        bash "add remote log server to centos config file" do
+          flags "-ex"
+          code <<-EOH
+            echo "\n\*.info @#{remote_server}:514\n\n" >> /etc/rsyslog.conf
+          EOH
+        end
```
Comment (douglaswth) [inline cookbooks/logging_rsyslog/providers/default.rb]: Please use an `only_if` block on the bash resource instead of doing the check at compile time.

File: cookbooks/sys_firewall/recipes/setup_rule.rb (hunk 1 above)
Comment (Neve) [inline cookbooks/sys_firewall/recipes/setup_rule.rb]: hmm
just do not think that we should remove this raise for now 
we can remove it when we will "upgrade" sys_firewall LWRP to use more complicated rules
because for now it put "all" directly in ipatables rule
-A FWR <% if @ip_addr -%>--source <%= @ip_addr %> <% end -%>
--protocol <%= @protocol %> --dport <%= @port %> -j ACCEPT
if we change "all" to "both" we will have to create two rules

Comment (douglaswth) [inline cookbooks/sys_firewall/recipes/setup_rule.rb]: From http://www.faqs.org/docs/iptables/matches.html#GENERICMATCHES:

> The command may also take a comma delimited list of protocols, such as udp,tcp which would match all UDP and TCP packets.

You can just pass `"udp,tcp"`.

**Comment by douglaswth** [comment]
Please give your pull requests and branches more descriptive names in the future.

File: cookbooks/sys_firewall/recipes/setup_rule.rb (hunk 1 above)
Comment (Neve) [inline cookbooks/sys_firewall/recipes/setup_rule.rb]: Yeah, that was the first thing i've tried to do
-A FWR -p tcp,udp --dport 8000 -j ACCEPT

 /usr/sbin/rebuild-iptables
iptables-restore v1.4.4: unknown protocol `tcp,udp' specified
Error occurred at line: 21

Comment (lopaka) [inline cookbooks/sys_firewall/recipes/setup_rule.rb]: "man iptables" shows you cannot put multiple protocols in the same command.  So, if input = both, should run 2 commands.  This is the recommended process.

File: cookbooks/sys_firewall/recipes/setup_rule.rb [hunk 3]
```diff
@@ -12,20 +12,29 @@
 raise "Invalid port specified: #{node[:sys_firewall][:rule][:port]}. Valid range 1-65536" unless rule_port > 0 and rule_port <= 65536
 rule_ip = node[:sys_firewall][:rule][:ip_address]
 rule_ip = (rule_ip == "" || rule_ip.downcase =~ /any/) ? nil : rule_ip
-rule_protocol = node[:sys_firewall][:rule][:protocol]
 to_enable = (node[:sys_firewall][:rule][:enable] == "enable") ? true : false
 
-if node[:sys_firewall][:enabled] == "enabled"
+# Create protocol array here to handle with "both" input
+if node[:sys_firewall][:rule][:protocol] == "both"
+rule_protocol = ["tcp","udp"]
```
Comment (lopaka) [inline cookbooks/sys_firewall/recipes/setup_rule.rb]: Please follow indent standards.

File: cookbooks/sys_firewall/recipes/setup_rule.rb [hunk 4]
```diff
@@ -12,20 +12,29 @@
...
+else
+rule_protocol = [node[:sys_firewall][:rule][:protocol]]
```
Comment (lopaka) [inline cookbooks/sys_firewall/recipes/setup_rule.rb]: Please follow indent standards.

File: cookbooks/logging_rsyslog/providers/default.rb [hunk 5]
```diff
@@ -60,12 +60,15 @@
...
       # /etc/rsyslog.d this should be removed.
+
+      # Skipping if entry already exists in /etc/rsyslog.conf
       log "  Configuring Redhat/CentOS."
       bash "add remote log server to centos config file" do
         flags "-ex"
         code <<-EOH
           echo "\n\*.info @#{remote_server}:514\n\n" >> /etc/rsyslog.conf
         EOH
+        not_if do ::File.open('/etc/rsyslog.conf', 'r') { |f| f.read }.include? "#{remote_server}" end
```
Comment (lopaka) [inline cookbooks/logging_rsyslog/providers/default.rb]: We should be searching for the whole line.  Perhaps creating a variable with the text of the rsyslog.conf entry and using that to add to the file and checking.

**Comment by lopaka** [comment]
Added comments that need addressing.

**Comment by lopaka** [comment]
Any reason why this has not been merged yet?
